src/APIController.react.js

```diff
@@ -49,6 +51,10 @@ class UnconnectedContainer extends Component {
             }
         }
 
+        if(hooks.request_pre !== null || hooks.request_post !== null) {
+            dispatch(setHooks(hooks));
```

Comment:
I believe that `dispatch` is asynchronous, so if we want to hydrate in this function (which makes sense, it's where we hydrate our API calls and our derived data), then we need to add a check below that `setHooks` has finished before we call `dispatch(hydrateInitialOutputs());`

Reply:
That is, in here: https://github.com/plotly/dash-renderer/pull/75/files#diff-1b2b6e62305d1ce2aaa047bb80a6f529R69

Reply:
The `dash-renderer` initialization is handled entirely in this logic:
1. The `layout` and `dependencies` API calls are made `dispatch(getLayout())`
2. The `paths` (`dispatch(computePaths({...}))` and `graphs` (`dispatch(computeGraphs(...))` stores are computed
3. Once that data is computed, we "hydrate" the initial outputs, that is we make the API calls to fire the callbacks for all of the outputs on the page `dispatch(hydrateInitialOutputs())`
4. Once everything is "hydrated", we render the app: ` else if (appLifecycle === APP_STATES('HYDRATED')) {`

These `dispatch` calls are asynchronous. So, when a `dispatch(...)` call is made, some stuff gets recomputed, and the component gets re-renderered, which fires this `initialization` call. This `initialization` call is repeatedly called until all of the stores are computed and the order in which stores are computed is described through the `if` statements. That is, this block:
```
        if (isEmpty(dependenciesRequest)) {
            dispatch(getDependencies());
        } else if (dependenciesRequest.status === 200 && isEmpty(graphs)) {
            dispatch(computeGraphs(dependenciesRequest.content));
}
```

is implicitly saying: first make the API call for the dependencies, _then_ compute the graphs based off the dependencies. The "_then_" is implicitly understood because of the lifecycle of the app: once the API call is finished, the component is _re-rendered_, firing this `initialization` function again (because `initialization` is fired in `componentWillReceiveProps`).

> As a side note, the alternative way to write this would be through a gnarly promise chain: `dispatch(getLayout()).then(dispatch(setLayout()).then(dispatch(computePaths(...))...` etc. I found that it's easier to reason about the order in which actions depend on each other (and the failure cases) through this more declarative representation where this function is just repeatedly called.

So, in this PR, you're introducing a new store. It's `null` by default and, since it affects the API calls, we need to make sure that it is computed before we start hydarting. You've set up the trigger to compute the hook, but you haven't set up any logic to prevent the hydration to happen before the hooks are finished. So, you need to describe that in https://github.com/plotly/dash-renderer/blob/bd3c8b38a23c4c953637d0518db00298f0f3d0c3/src/APIController.react.js#L58-L72, by adding another `if` statement that says "make sure that the hooks are set up before we start making the api calls". How you set up that if statement is up to you. In the other cases, we initialize the store to be `{}` or `null` and then we simply check if the store is _not_ `{}` or `null` (if it's not `{}` or `null`, then we know that it was computed). A similar check would probably work here too. Or, you could initialize the store to be some constent like `constants.EMPTY = 'EMPTY'` and then check if it's not `EMPTY`. But I think that `null` would work too.

I wouldn't use `status` or `200` here because those are used just for the API requests.